internal/cose.go

```diff
@@ -0,0 +1,165 @@
+package internal
+
+import (
+	"crypto/ecdsa"
+	"crypto/sha512"
+	"fmt"
+	"math/big"
+
+	"github.com/fxamacker/cbor/v2"
+)
+
+type SigningAlgorithm int
+
+const (
+	BadOrMissingAlgorithm SigningAlgorithm = iota
+	ES384
+)
+
+// https://datatracker.ietf.org/doc/html/rfc8152#section-4.2
+type CoseSign1 struct {
+	_ struct{} `cbor:",toarray"`
+
+	Protected   []byte
+	Unprotected cbor.RawMessage
+	Payload     []byte
+	Signature   []byte
+}
+
+type CoseHeader struct {
+	Alg interface{} `cbor:"1,keyasint,omitempty" json:"alg,omitempty"`
+}
+
+func MakeCoseSign1FromBytes(bytes []byte) (CoseSign1, error) {
+	coseSign1 := CoseSign1{}
+	err := cbor.Unmarshal(bytes, &coseSign1)
+	if nil != err {
+		return CoseSign1{}, fmt.Errorf("unmarshaling CoseSign1 from bytes: %w", err)
+	}
+
+	if len(coseSign1.Protected) == 0 {
+		return CoseSign1{}, fmt.Errorf("missing cose protected headers")
+	}
+
+	if len(coseSign1.Payload) == 0 {
+		return CoseSign1{}, fmt.Errorf("missing cose payload")
+	}
+
+	if len(coseSign1.Signature) == 0 {
+		return CoseSign1{}, fmt.Errorf("missing cose signature")
+	}
+	return coseSign1, nil
+}
+
+func (c CoseSign1) CheckSignature(
+	publicKey *ecdsa.PublicKey,
+) ([]byte, error) {
+	signingAlg, err := c.GetSigningAlgorithm()
+	if err != nil {
+		return nil, fmt.Errorf("extracting signing algorithm: %w", err)
+	}
```

Comment:
One change I made was to extract the signing algo from the CoseSign1 header and ensure that it matches the signing algorithm of the public key passed in (this is enforced by the switch statement on L84)